### PR TITLE
config, sysvar: map `max-server-connections` to `max_connections`

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: ref #34960

Problem Summary:

Before, if we want to change the limitation of `max-server-connections`, we have to change the configuration and restart the node. And the system variable `max_connections` is not available.

### What is changed and how it works?

Add a new config item `instance.max_connections` and make sysvar `max_connections` available.

### Check List

Tests 

- [x] Unit test (using previous testing)
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
> Thanks @morgo for suggesting this manual test!
```
Test 1: check whether 'max_connections' limits the number of incoming connections:
1. start a tidb-server
2. start one mysql client connecting to the server with unlimited 'max_connections'
3. 'set @@global.max_connections = 3'
4. try to start other mysql clients connecting to the server one by one, until some errors happen.
Result: only three clients can connect to the server.
The later connections encounter 'ERROR 1040 (08004): Too many connections'

Test 2: check whether 'set @@global.max_connections = ?' success:
1. start a tidb-server
2. start 5 mysql clients connecting to the server with unlimited 'max_connections'
3. try to 'set @@global.max_connections = ?' using the number from 0 to 10
Result: whatever the value is, the setting is always successful. 
But if the setting value is bigger than the current connection number, no other connections can come in.
This behaviour is the same with MySQL.
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility
> For [MySQL 8.0](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_connections), the range of `max_connections` is `[1, 100000]`, and the default value is `151`. In TiDB, since the default value of `instance.max_connections` is `0`, which means *no limitation*, so the range of `max_connections` in TiDB is `[0, 100000]`, and default value is `0`.

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by morgo — COMMENTED** [review]
Looks great! Can you manual test as well: what happens when you have 10 idle connections on a server, and then you set the max_connections to 5? I assume from the code it takes effect on the next connection, and the existing connections work fine.

For another test: what happens when you have 10 idle, and set the max_connections to 12? Do 2 new connections come in with no problems, but the 13th breaks?

**Comment by morgo** [comment]
It looks like the error returned if connections is exceeded is not the best :( This might be an existing bug - here is what MySQL returns:
```
$ mysql
ERROR 1040 (HY000): Too many connections
```

**Comment by CbcWestwolf** [comment]
> It looks like the error returned if connections is exceeded is not the best

Fixed it. Now the behavior is compatible with MySQL. PTAL :)

**Review by xiongjiwei — APPROVED** [review]

File: server/server.go [hunk 1]
```diff
@@ -514,11 +514,14 @@ func (s *Server) onConn(conn *clientConn) {
...
 		}
-		if errors.Cause(err) == io.EOF {
+		switch errors.Cause(err) {
+		case io.EOF:
 			// `EOF` means the connection is closed normally, we do not treat it as a noticeable error and log it in 'DEBUG' level.
 			logutil.BgLogger().With(zap.Uint64("conn", conn.connectionID)).
 				Debug("EOF", zap.String("remote addr", conn.bufReadConn.RemoteAddr().String()))
-		} else {
+		case errConCount:
+			_ = conn.writeError(ctx, err)
```
Comment (bb7133) [inline server/server.go:523]:
```suggestion
			_ = conn.writeError(ctx, err)
```
Please handle the error here, you could simply write an log for it.

Comment (CbcWestwolf) [reply server/server.go]: Fixed. PTAL :)

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**